src/iOS/Avalonia.iOS/Storage/IOSStorageProvider.cs

```diff
@@ -161,10 +161,77 @@ private async Task<IReadOnlyList<IStorageFile>> OpenDocumentsPickerAsync(FilePic
+            To save a file, we need to present the user with a document picker
+            This requires a temp file to be created and used to "export" the file to.
+            When the user picks the file location and name, UIDocumentPickerViewController
+            will give back the URI to the real file location, which we can then use 
+            to give back as an IStorageFile.
+            https://developer.apple.com/documentation/uikit/uidocumentpickerviewcontroller
+            Yes, it is weird, but without the temp file it will explode.
+        */
+
+        // Create a temporary file to use with the document picker
+        var tempFileName = options.SuggestedFileName ?? "document";
+        if (!string.IsNullOrEmpty(options.DefaultExtension))
+        {
+            var extension = options.DefaultExtension.StartsWith('.') ? options.DefaultExtension : "." + options.DefaultExtension;
+            if (!tempFileName.EndsWith(extension, StringComparison.OrdinalIgnoreCase))
+            {
+                tempFileName += extension;
+            }
+        }
+        
+        var tempDir = NSFileManager.DefaultManager.GetTemporaryDirectory().Append(Guid.NewGuid().ToString(), true);
+        if (tempDir == null)
+        {
+            throw new InvalidOperationException("Failed to get temporary directory for save file picker");
+        }
+
+        var isDirectoryCreated = NSFileManager.DefaultManager.CreateDirectory(tempDir, true, null, out var error);
+        if (!isDirectoryCreated)
+        {
+            throw new InvalidOperationException("Failed to create temporary directory for save file picker");
+        }
+        
+        var tempFileUrl = tempDir.Append(tempFileName, false);
+        
+        // Create an empty file at the temp location
+        NSData.FromString("").Save(tempFileUrl, false);
+        
+        UIDocumentPickerViewController documentPicker;
+        if (OperatingSystem.IsIOSVersionAtLeast(14))
+        {
+            documentPicker = new UIDocumentPickerViewController(new[] { tempFileUrl }, asCopy: true);
+        }
+        else
+        {
+#pragma warning disable CA1422
+            documentPicker = new UIDocumentPickerViewController(tempFileUrl, UIDocumentPickerMode.ExportToService);
+#pragma warning restore CA1422
+        }
+        
+        using (documentPicker)
+        {
+            if (OperatingSystem.IsIOSVersionAtLeast(13))
+            {
+                documentPicker.DirectoryUrl = GetUrlFromFolder(options.SuggestedStartLocation);
+            }
+            
+            documentPicker.Title = options.Title;
+            
+            var tcs = new TaskCompletionSource<NSUrl[]>();
+            documentPicker.Delegate = new PickerDelegate(urls => tcs.TrySetResult(urls));
+            var urls = await ShowPicker(documentPicker, tcs);
+            
+            // Clean up the temporary file
+            NSFileManager.DefaultManager.Remove(tempFileUrl, out _);
```

Comment:
In addition to the file, shouldn't the temporary directory be deleted?

Reply:
Updated with https://github.com/AvaloniaUI/Avalonia/pull/19364/commits/628e16be1ab8510a9b7d807ffccc0838c9cef113